server/service/src/apis/api_on_central/mod.rs

```diff
@@ -0,0 +1,122 @@
+use log::info;
+use repository::{NameStoreJoinRepository, NameStoreJoinRow, RepositoryError};
+use reqwest::{ClientBuilder, StatusCode};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use url::Url;
+use util::format_error;
+
+use crate::{
+    service_provider::{ServiceContext, ServiceProvider},
+    sync::CentralServerConfig,
+};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct NameStoreJoinParams {
+    pub id: String,
+    pub name_id: String,
+    pub store_id: String,
+}
+
+#[derive(Debug)]
+pub struct SiteAuth {
+    pub username: String,
+    pub password_sha256: String,
+}
+
+#[derive(Deserialize, Debug, Error, Serialize)]
+pub enum CentralApiError {
+    #[error("Not a central server")]
+    NotACentralServer,
+    #[error("Not authorized")]
+    NotAuthorized,
+    #[error("Connection error: {0}")]
+    ConnectionError(String),
+    #[error("Internal error: {0}")]
+    InternalError(String),
+}
+
+impl From<RepositoryError> for CentralApiError {
+    fn from(from: RepositoryError) -> Self {
+        CentralApiError::InternalError(format_error(&from))
+    }
+}
+
+/// Creates/updates a name_store_join for a patient
+pub async fn patient_name_store_join(
+    service_provider: &ServiceProvider,
+    NameStoreJoinParams {
+        id,
+        name_id,
+        store_id,
+    }: NameStoreJoinParams,
+    auth: SiteAuth,
+) -> Result<(), CentralApiError> {
+    if !CentralServerConfig::is_central_server() {
+        return Err(CentralApiError::NotACentralServer);
+    }
+
+    let ctx = service_provider.basic_context()?;
+    validate_auth(service_provider, &ctx, &auth).await?;
+
+    let name_store_join_repo = NameStoreJoinRepository::new(&ctx.connection);
+
+    // TODO: maybe should prevent this from creating a changelog? Let the OG one be source of truth?
+    name_store_join_repo.upsert_one(&NameStoreJoinRow {
+        id,
+        store_id: store_id.clone(),
+        // I think ideally would do a lookup and see if we have a name_link_id,
+        // but should do the same as in sync translation
+        name_link_id: name_id.clone(),
+
+        // This is only used for adding patient visibility, so ok to set these here
+        name_is_customer: true,
+        name_is_supplier: false,
+    })?;
+
+    info!(
+        "Created name_store_join for patient {} and store {}",
+        name_id, store_id
+    );
+
+    Ok(())
+}
+
+// OMS Central does not yet do auth validation for site credentials
+// So we call an OG endpoint to validate the credentials
+// (/login endpoint is for user credentials, using name_store_join for site credentials)
+async fn validate_auth(
```

Comment:
I don't love this, but couldn't think of better way, needed to be site auth which we don't support on OMS central yet. Also, set the API to accept the creds as Basic Auth, but @andreievg if you have thoughts about how we want to handle OMS central API auth please speak now 😁

Reply:
The same way we do in sync currently